<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-08-23">

<title>Cell type mapping benchmark</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="cell_type_mapping_benchmark_files/libs/clipboard/clipboard.min.js"></script>
<script src="cell_type_mapping_benchmark_files/libs/quarto-html/quarto.js"></script>
<script src="cell_type_mapping_benchmark_files/libs/quarto-html/popper.min.js"></script>
<script src="cell_type_mapping_benchmark_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="cell_type_mapping_benchmark_files/libs/quarto-html/anchor.min.js"></script>
<link href="cell_type_mapping_benchmark_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="cell_type_mapping_benchmark_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="cell_type_mapping_benchmark_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="cell_type_mapping_benchmark_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="cell_type_mapping_benchmark_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Cell type mapping benchmark</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 23, 2023</p>
    </div>
  </div>
    
  </div>
  

</header>

<p>Load required packages</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(semla)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">'TabulaMurisSenisData'</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"TabulaMurisSenisData"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TabulaMurisSenisData)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(spacexr)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pbapply)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load single-cell data (Allen Brain atlas) and Visium data (mouse brain tissue section).</p>
<section id="load-data" class="level2">
<h2 class="anchored" data-anchor-id="load-data">Load data</h2>
<p>We have access to 23 annotated cell types in the single-cell data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(se_allen, <span class="at">group.by =</span> <span class="st">"subclass"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/allen_brain_atlas.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">1. The reference single-cell data set contains 23 annotated cell types.</figcaption><p></p>
</figure>
</div>
</section>
<section id="create-dummy-expression-profiles" class="level2">
<h2 class="anchored" data-anchor-id="create-dummy-expression-profiles">Create dummy expression profiles</h2>
<p>Here we’ll create a synthetic Visium data set by sample single-cells from the Allen Brain atlas data set:</p>
<ol type="1">
<li>Downsample UMIs to 1%. This is to make sure that the synthetic data have roughly the same library sizes as the Visium mouse brain tissue section data set.</li>
<li>Calculate cell type weights. These weights will be used to determine the probability of a cell type being samples for the synthetic spots. Abundant cell type will be assigned higher probabilities and vice versa.</li>
<li>Sample cell numbers per synthetic spot. Here we’ll draw counts from a poisson distribution with the lambda (mean) parameter set to 10.</li>
<li>Select cell indices. The indices will be randomly selected using the cell type weights defined in step 2.</li>
<li>Aggregate expression vectors for selected indices. Each synthetic spot will consist of the averaged gene expression vectors obtained from the sampled cells.</li>
<li>Combine aggregated expression profiles to a final synthetic count matrix.</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>se_allen <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(se_allen, <span class="at">nfeatures =</span> <span class="fl">5e3</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>se_allen <span class="ot">&lt;-</span> se_allen[<span class="fu">VariableFeatures</span>(se_allen), ]</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>umis <span class="ot">&lt;-</span> <span class="fu">GetAssayData</span>(se_allen, <span class="at">slot =</span> <span class="st">"counts"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># 1. Down sample UMIs</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(se_allen), <span class="fu">cut</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(se_allen), <span class="at">breaks =</span> <span class="dv">50</span>))</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>batches <span class="ot">&lt;-</span> <span class="fu">pblapply</span>(bins, <span class="cf">function</span>(inds) {</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>  umis_subset <span class="ot">&lt;-</span> scuttle<span class="sc">::</span><span class="fu">downsampleMatrix</span>(umis[, inds], <span class="at">prop =</span> <span class="fl">0.01</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(umis_subset)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>}, <span class="at">cl =</span> <span class="dv">7</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>downsampled_umis <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, batches)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># 2. Calculate cell type weights</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> <span class="fu">table</span>(se_allen<span class="sc">$</span>subclass)</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> props<span class="sc">/</span><span class="fu">sum</span>(props) <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="co"># 3. Sample cells</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the average count</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>average_count <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the number of samples</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>num_samples <span class="ot">&lt;-</span> <span class="fl">1e4</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate samples from a Poisson distribution with the specified average</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>sampled_counts <span class="ot">&lt;-</span> <span class="fu">rpois</span>(num_samples, <span class="at">lambda =</span> average_count)</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>sampled_counts[sampled_counts <span class="sc">==</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a><span class="co"># 4. Select cell indices</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>label_matrix <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">names</span>(props), <span class="cf">function</span>(label) se_allen<span class="sc">$</span>subclass <span class="sc">==</span> label)</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>sampled_indices <span class="ot">&lt;-</span> <span class="fu">pblapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(sampled_counts), <span class="cf">function</span>(i) {</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="at">x =</span> <span class="fu">names</span>(props), <span class="at">size =</span> sampled_counts[i], <span class="at">prob =</span> props, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true" tabindex="-1"></a>  y <span class="ot">&lt;-</span> <span class="fu">table</span>(x)</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true" tabindex="-1"></a>  inds <span class="ot">&lt;-</span> <span class="fu">c</span>()</span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (lbl <span class="cf">in</span> <span class="fu">unique</span>(x)) {</span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true" tabindex="-1"></a>    inds <span class="ot">&lt;-</span> <span class="fu">c</span>(inds, <span class="fu">sample</span>(<span class="at">x =</span> <span class="fu">which</span>(label_matrix[, lbl]), <span class="at">size =</span> y[lbl]))</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(inds)</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true" tabindex="-1"></a>}, <span class="at">cl =</span> <span class="dv">7</span>)</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true" tabindex="-1"></a><span class="co"># 5. Aggregate expression vectors</span></span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true" tabindex="-1"></a>bins <span class="ot">&lt;-</span> <span class="fu">split</span>(<span class="dv">1</span><span class="sc">:</span>num_samples, <span class="fu">cut</span>(<span class="dv">1</span><span class="sc">:</span>num_samples, <span class="at">breaks =</span> <span class="dv">50</span>))</span>
<span id="cb3-43"><a href="#cb3-43" aria-hidden="true" tabindex="-1"></a>batches <span class="ot">&lt;-</span> <span class="fu">pblapply</span>(bins, <span class="cf">function</span>(inds) {</span>
<span id="cb3-44"><a href="#cb3-44" aria-hidden="true" tabindex="-1"></a>  umis_subset <span class="ot">&lt;-</span> downsampled_umis[, sampled_indices[inds] <span class="sc">|&gt;</span> <span class="fu">unlist</span>()]</span>
<span id="cb3-45"><a href="#cb3-45" aria-hidden="true" tabindex="-1"></a>  umis_subset <span class="ot">&lt;-</span> <span class="fu">rowsum</span>(umis_subset <span class="sc">|&gt;</span> <span class="fu">t</span>(), <span class="at">group =</span> <span class="fu">rep</span>(<span class="at">x =</span> inds, sampled_indices[inds] <span class="sc">|&gt;</span> <span class="fu">sapply</span>(length))) <span class="sc">|&gt;</span> <span class="fu">t</span>()</span>
<span id="cb3-46"><a href="#cb3-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(umis_subset)</span>
<span id="cb3-47"><a href="#cb3-47" aria-hidden="true" tabindex="-1"></a>}, <span class="at">cl =</span> <span class="dv">7</span>)</span>
<span id="cb3-48"><a href="#cb3-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-49"><a href="#cb3-49" aria-hidden="true" tabindex="-1"></a><span class="co"># 6. Combine aggregated expression profiles</span></span>
<span id="cb3-50"><a href="#cb3-50" aria-hidden="true" tabindex="-1"></a>mini_bulk_umis <span class="ot">&lt;-</span> <span class="fu">do.call</span>(cbind, batches)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="export-allen-brain-atlas-data-for-cell2location-and-stereoscope" class="level2">
<h2 class="anchored" data-anchor-id="export-allen-brain-atlas-data-for-cell2location-and-stereoscope">Export allen brain atlas data for cell2location and stereoscope</h2>
<p>Here we’ll export a subset of the single-cell data including 250 randomly selected cells per cell type. If a cell type has less than 250 cells, all cells will be selected. Only cell types with more than 10 cells are kept.</p>
<p>The exported matrices and tables will be used to run stereoscope and cell2location in a separate environment.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>seed <span class="ot">=</span> 1337L</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>nCells_per_group <span class="ot">&lt;-</span> <span class="dv">250</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(seed)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample barcodes</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>barcodes <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">barcode =</span> <span class="fu">colnames</span>(se_allen), <span class="at">group =</span> se_allen<span class="sc">$</span>subclass) <span class="sc">|&gt;</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(group) <span class="sc">|&gt;</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">slice</span>(<span class="fu">sample</span>(<span class="fu">min</span>(nCells_per_group, <span class="fu">n</span>())))</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove low abundant cell types</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>cells_per_celltype <span class="ot">&lt;-</span> <span class="fu">table</span>(barcodes<span class="sc">$</span>group)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">names</span>(cells_per_celltype)[cells_per_celltype <span class="sc">&gt;</span> <span class="dv">10</span>]</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>barcodes <span class="ot">&lt;-</span> barcodes <span class="sc">|&gt;</span> <span class="fu">filter</span>(group <span class="sc">%in%</span> keep)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co"># Export matrices and tables</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">"wynthetic_data"</span>)</span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">GetAssayData</span>(se_allen, <span class="at">slot =</span> <span class="st">"counts"</span>)[, barcodes<span class="sc">$</span>barcode] <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>(), <span class="at">file =</span> <span class="st">"synthetic_data/allen_brain_umis.tsv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a><span class="fu">writeLines</span>(<span class="fu">rownames</span>(se_allen), <span class="at">con =</span> <span class="st">"../../manuscript/synthetic_data/variable_genes.txt"</span>)</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(se_allen<span class="sc">@</span>meta.data[barcodes<span class="sc">$</span>barcode, ] <span class="sc">%&gt;%</span> <span class="fu">select</span>(subclass), <span class="at">file =</span> <span class="st">"../../manuscript/synthetic_data/metadata.tsv"</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>data.table<span class="sc">::</span><span class="fu">fwrite</span>(<span class="fu">GetAssayData</span>(se_Visium_agg, <span class="at">slot =</span> <span class="st">"counts"</span>) <span class="sc">%&gt;%</span> <span class="fu">t</span>() <span class="sc">%&gt;%</span> <span class="fu">as.data.frame</span>(), <span class="at">file =</span> <span class="st">"../../manuscript/synthetic_data/Visium_synthetic_spots_umis.tsv"</span>, <span class="at">quote =</span> <span class="cn">FALSE</span>, <span class="at">sep =</span> <span class="st">"</span><span class="sc">\t</span><span class="st">"</span>, <span class="at">row.names =</span> <span class="cn">TRUE</span>, <span class="at">col.names =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="create-a-seurat-object-from-the-synthetic-count-matrix" class="level3">
<h3 class="anchored" data-anchor-id="create-a-seurat-object-from-the-synthetic-count-matrix">Create a Seurat object from the synthetic count matrix</h3>
<p>Each synthetic spot will be named “barcode1”, barcode2”, …</p>
<p>For downstream steps, we need to normalize the data and select a set of top 5,000 variable genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mini_bulk_umis) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"barcode"</span>, <span class="dv">1</span><span class="sc">:</span>num_samples)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>se_Visium_agg <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> mini_bulk_umis, <span class="at">assay =</span> <span class="st">"Spatial"</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>se_Visium_agg <span class="ot">&lt;-</span> se_Visium_agg <span class="sc">|&gt;</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">5000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-nnls" class="level3">
<h3 class="anchored" data-anchor-id="run-nnls">Run NNLS</h3>
<p>The <code>RunNNLS()</code> method requires a <code>Seurat</code> object with normalized 10x Visium data and a <code>Seurat</code> object with normalized single-cell data. The <code>groups</code> argument defines what meta data column the cell type labels should be taken from in the single-cell <code>Seurat</code> object. In our single-cell <code>Seurat</code> object, the labels are stored in the “subclass” column.</p>
</section>
<section id="computation-time" class="level3">
<h3 class="anchored" data-anchor-id="computation-time">Computation time</h3>
<p>For the NNLS method, we’ll run 10 rounds, adding 10,000 spots to each round and starting with 10,000 spots. Below we create a “large” Seurat object with 100,000 spots by merging our <code>se_Visium_agg</code> object 10 times.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Merge Seurat object to have 100k spots</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>se_Visium_agg_large <span class="ot">&lt;-</span> <span class="fu">merge</span>(se_Visium_agg, <span class="at">y =</span> <span class="fu">list</span>(se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg, se_Visium_agg))</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_Visium_agg_large) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>se_Visium_agg_large <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(se_Visium_agg_large, <span class="at">nfeatures =</span> <span class="fl">5e3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Since the NNLS method runs in a matter of seconds, we’ll run each round in 10 iterations to compute average computation times.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>nnls_results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="fl">1e4</span>, <span class="fl">1e5</span>, <span class="fl">1e4</span>)) {</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">DefaultAssay</span>(se_Visium_agg_large) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  se_Visium_agg_subset <span class="ot">&lt;-</span> se_Visium_agg_large[, <span class="dv">1</span><span class="sc">:</span>i]</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  iters <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(n) {</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">cat</span>(<span class="fu">paste0</span>(<span class="st">"  iter: "</span>, n, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>))</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>    ti <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    tmp <span class="ot">&lt;-</span> <span class="fu">RunNNLS</span>(<span class="at">object =</span> se_Visium_agg_subset, <span class="at">nCells_per_group =</span> <span class="dv">250</span>, </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">singlecell_object =</span> se_allen, </span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">groups =</span> <span class="st">"subclass"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>    stamp <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>() <span class="sc">-</span> ti</span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(stamp)</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  nnls_results[[<span class="fu">paste0</span>(<span class="st">"n"</span>, i)]] <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">n =</span> i, <span class="at">stamp =</span> iters)</span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(se_Visium_agg_subset)</span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>nnls_stamps <span class="ot">&lt;-</span> <span class="fu">do.call</span>(bind_rows, nnls_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can plot the average computation times for each data set size.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(nnls_stamps <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(n) <span class="sc">%&gt;%</span> <span class="fu">summarize</span>(<span class="at">mean =</span> <span class="fu">mean</span>(stamp), <span class="at">sd =</span> <span class="fu">sd</span>(stamp)), </span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> n, <span class="at">y =</span> mean, <span class="at">ymin =</span> mean <span class="sc">-</span> sd, <span class="at">ymax =</span> mean <span class="sc">+</span> sd)) <span class="sc">+</span> </span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_errorbar</span>(<span class="at">width =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of spots"</span>, <span class="at">y =</span> <span class="st">"Seconds"</span>, <span class="at">title =</span> <span class="st">"Computation time with NNLS (5,000 genes)"</span>) <span class="sc">+</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">labels=</span><span class="cf">function</span>(x) <span class="fu">format</span>(x, <span class="at">big.mark =</span> <span class="st">","</span>, <span class="at">scientific =</span> <span class="cn">FALSE</span>), </span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                     <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="fl">1e4</span>, <span class="fl">1e5</span>, <span class="fl">1e4</span>)) <span class="sc">+</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">50</span>, <span class="at">hjust =</span> <span class="dv">1</span>), <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/NNLS_computation_time.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">2. Computation time for NNLS on synthetic Visium data. The x-axis shows the number of spots and the y-axis shows the run time in seconds.</figcaption><p></p>
</figure>
</div>
</section>
</section>
<section id="compute-performance-metrics" class="level2">
<h2 class="anchored" data-anchor-id="compute-performance-metrics">Compute performance metrics</h2>
<p>For the performance assessment, we’ll run the NNLS method on our 10,000 spots <code>se_Visium_agg</code> data set. We can get the expected counts by counting the number of cell types using our sampled indices.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate expected counts per spot</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>expected_counts <span class="ot">&lt;-</span> <span class="fu">do.call</span>(bind_rows, <span class="fu">pblapply</span>(<span class="fu">seq_along</span>(sampled_indices), <span class="cf">function</span>(i) {</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> se_allen<span class="sc">$</span>subclass[sampled_indices[[i]]] <span class="sc">|&gt;</span> <span class="fu">table</span>()</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">spot =</span> <span class="fu">paste0</span>(<span class="st">"barcode"</span>, i), <span class="at">celltype =</span> <span class="fu">names</span>(x), <span class="at">lbl =</span> x <span class="sc">|&gt;</span> <span class="fu">as.integer</span>())</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>}, <span class="at">cl =</span> <span class="dv">7</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Run NNLS for 1e4 spots</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_Visium_agg) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>se_Visium_agg <span class="ot">&lt;-</span>  <span class="fu">RunNNLS</span>(<span class="at">object =</span> se_Visium_agg, </span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">singlecell_object =</span> se_allen, </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">groups =</span> <span class="st">"subclass"</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">min_prop =</span> <span class="dv">0</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># cast to wide format</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> expected_counts <span class="sc">|&gt;</span> tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> spot, <span class="at">names_from =</span> celltype, <span class="at">values_from =</span> lbl) <span class="sc">|&gt;</span> </span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"spot"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.matrix</span>()</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>props[<span class="fu">is.na</span>(props)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(props, <span class="at">margin =</span> <span class="dv">1</span>)</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> props[, <span class="fu">colnames</span>(props) <span class="sc">!=</span> <span class="st">"CR"</span>]</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>props <span class="ot">&lt;-</span> props[, <span class="fu">rownames</span>(se_Visium_agg)]</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare proportions</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>pred_all <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">inf_props_nnls =</span> <span class="fu">GetAssayData</span>(se_Visium_agg, <span class="at">slot =</span> <span class="st">"data"</span>) <span class="sc">|&gt;</span> <span class="fu">t</span>() <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>(),</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>             <span class="at">exp_props =</span> props <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>(), </span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>             <span class="at">celltype =</span> <span class="fu">rep</span>(<span class="fu">rownames</span>(se_Visium_agg), <span class="at">each =</span> <span class="fl">1e4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-rctd" class="level1">
<h1>Run RCTD</h1>
<p>Now we can run the RCTD method on our 10,000 spots data set and time the computation. The RCTD method could not be run locally for larger data sets and we therefore only ran the deconvolution for 10,00 spots. (Took ~19 minutes to run on a Macbook Pro 2017, 3.1 GHz Quad-Core Intel Core i7, 16GB).</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set up reference</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(se_allen) <span class="ot">&lt;-</span> <span class="st">"subclass"</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># extract information to pass to the RCTD Reference function</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>celltypes <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">/"</span>, <span class="at">replacement =</span> <span class="st">"."</span>, <span class="at">x =</span> se_allen<span class="sc">$</span>subclass)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>celltypes <span class="ot">&lt;-</span> celltypes[celltypes <span class="sc">!=</span> <span class="st">"CR"</span>]</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>cluster <span class="ot">&lt;-</span> <span class="fu">as.factor</span>(<span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">/"</span>, <span class="at">replacement =</span> <span class="st">"."</span>, <span class="at">x =</span> celltypes))</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>nUMI <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">GetAssayData</span>(se_allen, <span class="at">slot =</span> <span class="st">"counts"</span>))</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>reference <span class="ot">&lt;-</span> <span class="fu">Reference</span>(<span class="fu">GetAssayData</span>(se_allen, <span class="at">slot =</span> <span class="st">"counts"</span>)[, barcodes<span class="sc">$</span>barcode], cluster[barcodes<span class="sc">$</span>barcode], nUMI[barcodes<span class="sc">$</span>barcode])</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># set up query with the RCTD function SpatialRNA</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(mini_bulk_umis) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"barcode"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(mini_bulk_umis))</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="fl">1e4</span>, <span class="at">y =</span> <span class="dv">1</span><span class="sc">:</span><span class="fl">1e4</span>, <span class="at">row.names =</span> <span class="fu">paste0</span>(<span class="st">"barcode"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fl">1e4</span>))</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>query <span class="ot">&lt;-</span> <span class="fu">SpatialRNA</span>(<span class="at">coords =</span> coords, <span class="at">counts =</span> mini_bulk_umis, <span class="at">nUMI =</span> <span class="fu">colSums</span>(mini_bulk_umis))</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Run RCTD</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>ti <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">create.RCTD</span>(query, reference, <span class="at">max_cores =</span> <span class="dv">7</span>)</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>RCTD <span class="ot">&lt;-</span> <span class="fu">run.RCTD</span>(RCTD, <span class="at">doublet_mode =</span> <span class="st">'full'</span>)</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>rctd_stamp <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>() <span class="sc">-</span> ti</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="compute-performance-metrics-1" class="level2">
<h2 class="anchored" data-anchor-id="compute-performance-metrics-1">Compute performance metrics</h2>
<p>The results of RCTD full mode are stored in <code>@results$weights</code>. To obtain proportion estimates, we normalize the weights using <code>normalize_weights</code> so that they sum to one. Each entry represents the estimated proportion of each cell type on each spot.</p>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>barcodes <span class="ot">&lt;-</span> <span class="fu">colnames</span>(RCTD<span class="sc">@</span>spatialRNA<span class="sc">@</span>counts)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>weights <span class="ot">&lt;-</span> RCTD<span class="sc">@</span>results<span class="sc">$</span>weights</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>norm_weights <span class="ot">&lt;-</span> <span class="fu">normalize_weights</span>(weights)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(norm_weights) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">."</span>, <span class="at">replacement =</span> <span class="st">"/"</span>, <span class="at">x =</span> <span class="fu">colnames</span>(norm_weights))</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare proportions</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>pred_all<span class="sc">$</span>inf_props_rctd <span class="ot">&lt;-</span> norm_weights[, <span class="fu">colnames</span>(props)] <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="seurat" class="level1">
<h1>Seurat</h1>
<p>Finally, we’ll run the label transfer method available from <code>Seurat</code>. The method returns prediction scores for each cell type and spot which cannot be directly interpreted as proportions. However, if we normalize the prediction scores we can use them as a proxy for cell type proportions to calculate our performance metrics.</p>
<section id="computation-time-1" class="level2">
<h2 class="anchored" data-anchor-id="computation-time-1">Computation time</h2>
<p>The Seurat method runs relatively fast and we therefore logged the computation time running on synthetic Visium data from 10,000 to 100,000 spots.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>se_Visium_agg_large <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(se_Visium_agg_large, <span class="at">value =</span> <span class="st">"Spatial"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>se_Visium_agg_large <span class="ot">&lt;-</span> se_Visium_agg_large <span class="sc">|&gt;</span> <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span> <span class="fu">RunPCA</span>()</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>seurat_results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="fl">1e4</span>, <span class="fl">1e5</span>, <span class="fl">1e4</span>)) {</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(i)</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">DefaultAssay</span>(se_Visium_agg_large) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>  se_Visium_agg_subset <span class="ot">&lt;-</span> se_Visium_agg_large[, <span class="dv">1</span><span class="sc">:</span>i]</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  ti <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> se_allen, </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">query =</span> se_Visium_agg_subset)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  predictions.assay <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> anchors, </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">refdata =</span> se_allen<span class="sc">$</span>subclass, </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">prediction.assay =</span> <span class="cn">TRUE</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">weight.reduction =</span> se_Visium_agg_subset[[<span class="st">"pca"</span>]], </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  stamp <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>() <span class="sc">-</span> ti</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  seurat_results[[<span class="fu">paste0</span>(<span class="st">"n"</span>, i)]] <span class="ot">&lt;-</span> <span class="fu">tibble</span>(<span class="at">n =</span> i, <span class="at">stamp =</span> stamp)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>seurat_stamps <span class="ot">&lt;-</span> <span class="fu">do.call</span>(bind_rows, seurat_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Below we compare the run time for the NNLS and Seurat methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>gg <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(nnls_stamps <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">method =</span> <span class="st">"nnls"</span>) <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">stamp =</span> <span class="fu">as.numeric</span>(stamp)),</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>                seurat_stamps <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">method =</span> <span class="st">"Seurat"</span>) <span class="sc">|&gt;</span> <span class="fu">mutate</span>(<span class="at">stamp =</span> <span class="fu">as.numeric</span>(stamp)<span class="sc">*</span><span class="dv">60</span>))</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(gg <span class="sc">|&gt;</span> <span class="fu">group_by</span>(n, method) <span class="sc">|&gt;</span> <span class="fu">summarize</span>(<span class="at">mean =</span> <span class="fu">mean</span>(stamp), <span class="at">sd =</span> <span class="fu">sd</span>(stamp)), </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x =</span> n, <span class="at">y =</span> mean, <span class="at">ymin =</span> mean <span class="sc">-</span> sd, <span class="at">ymax =</span> mean <span class="sc">+</span> sd, <span class="at">color =</span> method)) <span class="sc">+</span> </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>() <span class="sc">+</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Number of spots"</span>, <span class="at">y =</span> <span class="st">"Seconds"</span>, <span class="at">title =</span> <span class="st">"Computation time with NNLS and Seurat (5,000 genes)"</span>) <span class="sc">+</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">labels=</span><span class="cf">function</span>(x) <span class="fu">format</span>(x, <span class="at">big.mark =</span> <span class="st">","</span>, <span class="at">scientific =</span> <span class="cn">FALSE</span>), </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>                     <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="fl">1e4</span>, <span class="fl">1e5</span>, <span class="fl">1e4</span>)) <span class="sc">+</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">annotation_logticks</span>() <span class="sc">+</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">50</span>, <span class="at">hjust =</span> <span class="dv">1</span>), <span class="at">panel.border =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="cn">NA</span>, <span class="at">colour =</span> <span class="st">"black"</span>))</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/NNLS_vs_Seurat_computation_time.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">3. Computation time for NNLS and Seurat label transfer on synthetic Visium data. The x-axis shows the number of spots and the y-axis shows the run time in seconds (log10 scale).</figcaption><p></p>
</figure>
</div>
</section>
<section id="compute-performance-metrics-2" class="level2">
<h2 class="anchored" data-anchor-id="compute-performance-metrics-2">Compute performance metrics</h2>
<p>Again, we can run the Seurat method on our 10,000 spots data set and normalize the results scores to obtain proportion estimates and calculate performance metrics.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>se_Visium_agg <span class="ot">&lt;-</span> <span class="fu">SetIdent</span>(se_Visium_agg, <span class="at">value =</span> <span class="st">"Spatial"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>se_Visium_agg <span class="ot">&lt;-</span> se_Visium_agg <span class="sc">|&gt;</span> <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span> <span class="fu">RunPCA</span>()</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>anchors <span class="ot">&lt;-</span> <span class="fu">FindTransferAnchors</span>(<span class="at">reference =</span> se_allen[, se_allen<span class="sc">$</span>subclass <span class="sc">!=</span> <span class="st">"CR"</span>], </span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>                               <span class="at">query =</span> se_Visium_agg)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>predictions.assay <span class="ot">&lt;-</span> <span class="fu">TransferData</span>(<span class="at">anchorset =</span> anchors, </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">refdata =</span> se_allen<span class="sc">$</span>subclass[se_allen<span class="sc">$</span>subclass <span class="sc">!=</span> <span class="st">"CR"</span>], </span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">prediction.assay =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">weight.reduction =</span> se_Visium_agg[[<span class="st">"pca"</span>]], </span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>predictions.assay.props <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(predictions.assay<span class="sc">@</span>data[<span class="fu">rownames</span>(predictions.assay) <span class="sc">!=</span> <span class="st">"max"</span>, ], <span class="at">margin =</span> <span class="dv">2</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare proportions</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>pred_all<span class="sc">$</span>inf_props_seurat <span class="ot">&lt;-</span> predictions.assay.props[<span class="fu">rownames</span>(se_Visium_agg<span class="sc">@</span>assays<span class="sc">$</span>celltypeprops), ] <span class="sc">|&gt;</span> <span class="fu">t</span>() <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="stereoscope" class="level1">
<h1>Stereoscope</h1>
<p>For deconvolution with stereoscope, we used the scvi implementation, following the tutorial <a href="https://docs.scvi-tools.org/en/stable/tutorials/notebooks/spatial/stereoscope_heart_LV_tutorial.html">‘STereoscope applied to left ventricule data’</a>. First, a model was trained on the scRNA-seq data with 1,000 epochs. Next, the model was used to deconvolve the synthetic Visium expression profiles in 2,000 epochs. For the stereoscope deconvolution, we used a NVIDIA A100-SMX4-80GB Tensor core GPU.</p>
<section id="compute-performance-metrics-3" class="level2">
<h2 class="anchored" data-anchor-id="compute-performance-metrics-3">Compute performance metrics</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>stereoscope <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"synthetic_data/stereoscope/stereoscope_proportions.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="dv">1</span>, <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>pred_all<span class="sc">$</span>inf_props_stereoscope <span class="ot">&lt;-</span> stereoscope[, <span class="fu">rownames</span>(se_Visium_agg<span class="sc">@</span>assays<span class="sc">$</span>celltypeprops)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="cell2location" class="level1">
<h1>Cell2location</h1>
<p>For deconvolution with cell2location, we followed the tutorial <a href="https://cell2location.readthedocs.io/en/latest/notebooks/cell2location_tutorial.html#Loading-Visium-and-scRNA-seq-reference-data">‘Mapping human lymph node cell types to 10X Visium with Cell2location’</a>. First, a model was trained on the scRNA-seq data with 1,000 epochs. Next, the model was used to deconvolve the synthetic Visium expression profiles in 30,000 epochs. For the cell2location deconvolution, we used a NVIDIA A100-SMX4-80GB Tensor core GPU.</p>
<section id="compute-performance-metrics-4" class="level2">
<h2 class="anchored" data-anchor-id="compute-performance-metrics-4">Compute performance metrics</h2>
<p>Load cell2location cell abundances.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>cell2location <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"synthetic_data/cell2location/q05_cell_abundance_w_sf.csv"</span>, <span class="at">sep =</span> <span class="st">","</span>, <span class="at">header =</span> <span class="cn">TRUE</span>, <span class="at">row.names =</span> <span class="dv">1</span>, <span class="at">check.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>cell2location_props <span class="ot">&lt;-</span> <span class="fu">prop.table</span>(cell2location <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>(), <span class="at">margin =</span> <span class="dv">1</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(cell2location_props) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"q05cell_abundance_w_sf_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> <span class="fu">colnames</span>(cell2location_props))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>pred_all<span class="sc">$</span>inf_props_cell2location <span class="ot">&lt;-</span> cell2location_props[, <span class="fu">rownames</span>(se_Visium_agg<span class="sc">@</span>assays<span class="sc">$</span>celltypeprops)] <span class="sc">|&gt;</span> <span class="fu">as.matrix</span>() <span class="sc">|&gt;</span> <span class="fu">as.numeric</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="run-time-for-10k-spots" class="level1">
<h1>Run time for 10k spots</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(run_time, <span class="fu">aes</span>(method, time)) <span class="sc">+</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_label</span>(<span class="fu">aes</span>(method, time, <span class="at">label =</span> time_format)) <span class="sc">+</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Deconvolution method"</span>, <span class="at">y =</span> <span class="st">"time in seconds"</span>, <span class="at">title =</span> <span class="st">"Run time for 10,000 spots"</span>)</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/run_time_comparison_10k.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">4. Computation time for 5 methods on synthetic Visium data with 10,000 spots. The y-axis shows the run time in seconds.</figcaption><p></p>
</figure>
</div>
<div class="cell" data-fig.asp="0.5">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pred_all_long <span class="ot">&lt;-</span> pred_all <span class="sc">|&gt;</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"inf_props_nnls"</span>, <span class="st">"inf_props_rctd"</span>, <span class="st">"inf_props_seurat"</span>, </span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>                               <span class="st">"inf_props_stereoscope"</span>, <span class="st">"inf_props_cell2location"</span>), </span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">names_to =</span> <span class="st">"method"</span>, <span class="at">values_to =</span> <span class="st">"inf_props"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"inf_props_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> method)) <span class="sc">|&gt;</span> </span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">factor</span>(method, <span class="at">levels =</span> <span class="fu">c</span>(<span class="st">"nnls"</span>, <span class="st">"rctd"</span>, <span class="st">"stereoscope"</span>, <span class="st">"cell2location"</span>, <span class="st">"seurat"</span>))) <span class="sc">|&gt;</span> </span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(celltype, method) <span class="sc">|&gt;</span> </span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">cor =</span> <span class="fu">cor</span>(exp_props, inf_props),</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>            <span class="at">rmse =</span> <span class="fu">rmse</span>(exp_props, inf_props), <span class="at">.groups =</span> <span class="st">"drop"</span>)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pred_all_long, <span class="fu">aes</span>(method, cor)) <span class="sc">+</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Pearson correlation (inferred vs expected proportions)"</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pred_all_long, <span class="fu">aes</span>(method, rmse)) <span class="sc">+</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_violin</span>(<span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"RMSE (inferred vs expected proportions)"</span>)</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/performance_metrics.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">5. Performance metrics for 5 methods. The pearson correlation scores and RMSE are calculated from the inferred vs expected proportions.</figcaption><p></p>
</figure>
</div>
<section id="scatter-plots" class="level2">
<h2 class="anchored" data-anchor-id="scatter-plots">Scatter plots</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pred_all_long <span class="ot">&lt;-</span> pred_all <span class="sc">|&gt;</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">c</span>(<span class="st">"inf_props_nnls"</span>, <span class="st">"inf_props_rctd"</span>, <span class="st">"inf_props_seurat"</span>, </span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>                                                <span class="st">"inf_props_stereoscope"</span>, <span class="st">"inf_props_cell2location"</span>), </span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                                       <span class="at">names_to =</span> <span class="st">"method"</span>, <span class="at">values_to =</span> <span class="st">"inf_props"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>         <span class="fu">mutate</span>(<span class="at">method =</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"inf_props_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="at">x =</span> method))</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>p <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(pred_all_long, </span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(exp_props, inf_props)) <span class="sc">+</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(celltype<span class="sc">~</span>method) <span class="sc">+</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm) <span class="sc">+</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Expected proportions"</span>, <span class="at">y =</span> <span class="st">"Inferred proportions"</span>)</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="synthetic_spots/scatter_plots_comparison.jpg" class="img-fluid figure-img"></p>
<p></p><figcaption aria-hidden="true" class="figure-caption">6. Scatter plot showing expected vs inferred cell type porportions for 5 methods on 10,000 synthetic Visium spots. The blue line indicates a fitter trend line and the dashed red lines indicate the expected 1 to 1 relationship.</figcaption><p></p>
</figure>
</div>
<p><br></p>
<hr>
<details open="">
<summary>
<strong>Package versions</strong>
</summary>
<ul>
<li><p><code>semla</code>: 1.1.0</p></li>
<li><p><code>RcppML</code>: 0.3.7</p></li>
</ul>
</details>
<details>
<summary>
<strong>Session info</strong>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.1 (2022-06-23)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] digest_0.6.29     lifecycle_1.0.3   jsonlite_1.8.3    magrittr_2.0.3   
 [5] evaluate_0.16     rlang_1.1.1       stringi_1.7.8     cli_3.4.1        
 [9] rstudioapi_0.14   vctrs_0.6.3       rmarkdown_2.15    tools_4.2.1      
[13] stringr_1.5.0     glue_1.6.2        htmlwidgets_1.5.4 xfun_0.32        
[17] yaml_2.3.5        fastmap_1.1.0     compiler_4.2.1    htmltools_0.5.4  
[21] knitr_1.39       </code></pre>
</div>
</div>
</details>
<p><br></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>