<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-08-31">

<title>Cell type mapping with NNLS</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="cell_type_mapping_with_NNLS_files/libs/clipboard/clipboard.min.js"></script>
<script src="cell_type_mapping_with_NNLS_files/libs/quarto-html/quarto.js"></script>
<script src="cell_type_mapping_with_NNLS_files/libs/quarto-html/popper.min.js"></script>
<script src="cell_type_mapping_with_NNLS_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="cell_type_mapping_with_NNLS_files/libs/quarto-html/anchor.min.js"></script>
<link href="cell_type_mapping_with_NNLS_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="cell_type_mapping_with_NNLS_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="cell_type_mapping_with_NNLS_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="cell_type_mapping_with_NNLS_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="cell_type_mapping_with_NNLS_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Cell type mapping with NNLS</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 31, 2023</p>
    </div>
  </div>
    
  </div>
  

</header>

<p>Cell type mapping with SRT data refers to a set of methods which allows you to infer the quantity of cells from SRT expression profiles. <code>semla</code> offers a quick method based on <em>Non-Negative Least Squares</em> to infer cell type proportions directly from Visium spot expression profiles. We utilize the NNLS method implemented in the <a href="https://github.com/zdebruine/RcppML">RcppML</a> <em>v0.3.7</em> package available on CRAN: <a href="https://cran.r-project.org/web/packages/RcppML/index.html" class="uri">https://cran.r-project.org/web/packages/RcppML/index.html</a>. Please ensure you have the correct version installed before proceeding with this tutorial.</p>
<p>In this tutorial, we’ll take a look at two examples. First, we have two tissue sections of a <a href="#mbrain">mouse brain</a> and a single cell RNA-seq dataset from the Allen Brain Atlas. In the second example, we have a <a href="#mkidney">mouse kidney</a> section and single cell RNA-seq data from the Tabula Muris Senis.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(semla)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">'TabulaMurisSenisData'</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)) {</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"TabulaMurisSenisData"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(TabulaMurisSenisData)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(purrr)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="mbrain" class="level2">
<h2 class="anchored" data-anchor-id="mbrain">Mouse brain</h2>
<p>For this tutorial, we will use a single-cell data set from the Allen Brain Atlas. You can check the code chunk below or follow <a href="https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1">this</a> link to download it manually.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">timeout=</span><span class="dv">200</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>tmpdir <span class="ot">&lt;-</span> <span class="st">"."</span> <span class="co"># Set current wd or change to tmpdir()</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(tmpdir, <span class="st">"/mousebrain"</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>targetdir <span class="ot">&lt;-</span> <span class="fu">paste0</span>(tmpdir, <span class="st">"/mousebrain"</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/single-cell"</span>))</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>destfile <span class="ot">&lt;-</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/single-cell/allen_brain.rds"</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="st">"https://www.dropbox.com/s/cuowvm4vrf65pvq/allen_cortex.rds?dl=1"</span>, <span class="at">destfile =</span> destfile)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We will also need to download the 10x Visium data from 10x Genomics website. You can download the files directly with R by following the code chunk below or download the data directly from <a href="https://www.10xgenomics.com/resources/datasets">here</a>.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium"</span>))</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Download section 1</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1"</span>))</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_filtered_feature_bc_matrix.h5"</span>, </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1/filtered_feature_bc_matrix.h5"</span>))</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Anterior/V1_Mouse_Brain_Sagittal_Anterior_spatial.tar.gz"</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1/spatial.tar.gz"</span>))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a><span class="fu">untar</span>(<span class="at">tarfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1/spatial.tar.gz"</span>), </span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">exdir =</span>  <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1/"</span>))</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">file.remove</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S1/spatial.tar.gz"</span>))</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Download section 2</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2"</span>))</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_filtered_feature_bc_matrix.h5"</span>, </span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2/filtered_feature_bc_matrix.h5"</span>))</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.0.0/V1_Mouse_Brain_Sagittal_Posterior/V1_Mouse_Brain_Sagittal_Posterior_spatial.tar.gz"</span>,</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2/spatial.tar.gz"</span>))</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a><span class="fu">untar</span>(<span class="at">tarfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2/spatial.tar.gz"</span>), </span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>      <span class="at">exdir =</span>  <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2/"</span>))</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a><span class="fu">file.remove</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium/S2/spatial.tar.gz"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Load Visium data with semla into a Seurat object. The following steps assumes that the mouse brain 10x Visium data is located in <code>./mousebrain/visium/</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assemble spaceranger output files</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="st">"./mousebrain/visium/*/filtered_feature_bc_matrix.h5"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>imgs <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="st">"./mousebrain/visium/*/spatial/tissue_hires_image.png"</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>spotfiles <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="st">"./mousebrain/visium/*/spatial/tissue_positions_list.csv"</span>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>json <span class="ot">&lt;-</span> <span class="fu">Sys.glob</span>(<span class="st">"./mousebrain/visium/*/spatial/scalefactors_json.json"</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>infoTable <span class="ot">&lt;-</span> <span class="fu">tibble</span>(samples, imgs, spotfiles, json, </span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">section_id =</span> <span class="fu">paste0</span>(<span class="st">"section_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>))</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Seurat object with 1 Sagittal Anterior section and 1 Sagittal Posterior section</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>se_brain_spatial <span class="ot">&lt;-</span> <span class="fu">ReadVisiumData</span>(infoTable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Load single-cell data</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>se_allen <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"./mousebrain/single-cell/allen_brain.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="normalize-data" class="level3">
<h3 class="anchored" data-anchor-id="normalize-data">Normalize data</h3>
<p>Here we will apply the same log-normalization procedure to both the 10x Visium data (<code>se_brain_spatial</code>) and to the single-cell data (<code>se_brain_singlecell</code>). We set the number of variable features quite high because later on we will use the intersect between the variable features in the single-cell data and the variable features in the 10x Visium data for NNLS. The NNLS method is quite fast so there is actually no need to select only a subset of features. Instead, we can just use all genes that are shared across the single-cell and 10x Visium data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data and find variable features for Visium data</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>se_brain_spatial <span class="ot">&lt;-</span> se_brain_spatial <span class="sc">|&gt;</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">10000</span>)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalize data and run vanilla analysis to create UMAP embedding</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>se_allen <span class="ot">&lt;-</span> se_allen <span class="sc">|&gt;</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>() <span class="sc">|&gt;</span> </span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span> </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunPCA</span>() <span class="sc">|&gt;</span> </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Rerun FindVariableFeatures to increase the number before cell type deconvolution</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>se_allen <span class="ot">&lt;-</span> se_allen <span class="sc">|&gt;</span> </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can visualize the available cell types in our UMAP embedding of the cells. We have access to 23 annotated cell types, including L2-L6 layer neurons which have a distinct spatial distribution in the tissue.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DimPlot</span>(se_allen, <span class="at">group.by =</span> <span class="st">"subclass"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-10-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="run-nnls" class="level3">
<h3 class="anchored" data-anchor-id="run-nnls">Run NNLS</h3>
<p>The <code>RunNNLS()</code> method requires a normalized <code>Seurat</code> object with 10x Visium data and a a normalized <code>Seurat</code> object with single-cell data. The <code>groups</code> argument defines where the cell type labels should be taken from in the single-cell <code>Seurat</code> object. In our single-cell <code>Seurat</code> object, the labels are stored in the “subclass” column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_brain_spatial) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>ti <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>se_brain_spatial <span class="ot">&lt;-</span> <span class="fu">RunNNLS</span>(<span class="at">object =</span> se_brain_spatial, </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                            <span class="at">singlecell_object =</span> se_allen, </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                            <span class="at">groups =</span> <span class="st">"subclass"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Predicting cell type proportions ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Fetching data from Seurat objects</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Filtering out features that are only present in one data set</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Kept 3378 features for deconvolution</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required namespace: RcppML</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Preparing data for NNLS</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Downsampling scRNA-seq data to include a maximum of 50 cells per cell type</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Cell type(s) CR have too few cells (&lt;10) and will be excluded</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Kept 22 cell types after filtering</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Calculating cell type expression profiles</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Predicting cell type proportions with NNLS for 22 cell types</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Returning results in a new 'Assay' named 'celltypeprops'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Setting default assay to 'celltypeprops'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ Finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sprintf</span>(<span class="st">"RunNNLS completed in %s seconds"</span>, <span class="fu">round</span>(<span class="fu">Sys.time</span>() <span class="sc">-</span> ti, <span class="at">digits =</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "RunNNLS completed in 2.56 seconds"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check available cell types</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(se_brain_spatial)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Astro"      "Endo"       "L2/3 IT"    "L4"         "L5 IT"     
 [6] "L5 PT"      "L6 CT"      "L6 IT"      "L6b"        "Lamp5"     
[11] "Macrophage" "Meis2"      "NP"         "Oligo"      "Peri"      
[16] "Pvalb"      "SMC"        "Serpinf1"   "Sncg"       "Sst"       
[21] "VLMC"       "Vip"       </code></pre>
</div>
</div>
<p>NB: the CR cell type was discarded because the number of cell was lower than 10. 10 is the lower limit for the allowed number of cells per cell type but this can be overridden with <code>minCells_per_celltype</code>.</p>
<p>The plots below show the spatial distributions of proportions for a selected set of cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r fold-hide code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot selected cell types</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_brain_spatial) <span class="ot">&lt;-</span> <span class="st">"celltypeprops"</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>selected_celltypes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"L2/3 IT"</span>, <span class="st">"L4"</span>, <span class="st">"L5 IT"</span>, </span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>                       <span class="st">"L5 PT"</span>, <span class="st">"L6 CT"</span>, <span class="st">"L6 IT"</span>, <span class="st">"L6b"</span>, </span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>                       <span class="st">"Oligo"</span>, <span class="st">"Pvalb"</span>, <span class="st">"Meis2"</span>, <span class="st">"Astro"</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>                       <span class="st">"VLMC"</span>, <span class="st">"SMC"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>se_brain_spatial <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(se_brain_spatial, <span class="at">image_height =</span> <span class="fl">1e3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Loading H&amp;E images ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Loading image from ./mousebrain/visium/S1/spatial/tissue_hires_image.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Scaled image from 1998x2000 to 1000x1001 pixels</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Loading image from ./mousebrain/visium/S2/spatial/tissue_hires_image.png</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Scaled image from 1998x2000 to 1000x1001 pixels</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Saving loaded H&amp;E images as 'rasters' in Seurat object</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(selected_celltypes), <span class="cf">function</span>(i) {</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">MapFeatures</span>(se_brain_spatial, <span class="at">pt_size =</span> <span class="fl">1.3</span>,</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> selected_celltypes[i], <span class="at">image_use =</span> <span class="st">"raw"</span>,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">arrange_features =</span> <span class="st">"row"</span>, <span class="at">scale =</span> <span class="st">"shared"</span>, </span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>,</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">colors =</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">9</span>, <span class="at">name =</span> <span class="st">"Spectral"</span>) <span class="sc">|&gt;</span> <span class="fu">rev</span>(), </span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>)  <span class="sc">+</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">&amp;</span> </span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>, <span class="at">legend.margin =</span> <span class="fu">margin</span>(<span class="at">b =</span> <span class="dv">50</span>),</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">0</span>),</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">setNames</span>(<span class="at">nm =</span> selected_celltypes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="mapped-mouse-brain-cell-types" class="level3 tabset">
<h3 class="tabset anchored" data-anchor-id="mapped-mouse-brain-cell-types">Mapped mouse brain cell types</h3>
<section id="l23-it" class="level4">
<h4 class="anchored" data-anchor-id="l23-it">L2/3 IT</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l4" class="level4">
<h4 class="anchored" data-anchor-id="l4">L4</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-2.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l5-it" class="level4">
<h4 class="anchored" data-anchor-id="l5-it">L5 IT</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-3.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l5-pt" class="level4">
<h4 class="anchored" data-anchor-id="l5-pt">L5 PT</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-4.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l6-ct" class="level4">
<h4 class="anchored" data-anchor-id="l6-ct">L6 CT</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-5.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l6-it" class="level4">
<h4 class="anchored" data-anchor-id="l6-it">L6 IT</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-6.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="l6b" class="level4">
<h4 class="anchored" data-anchor-id="l6b">L6b</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-7.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="oligo" class="level4">
<h4 class="anchored" data-anchor-id="oligo">Oligo</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-8.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="pvalb" class="level4">
<h4 class="anchored" data-anchor-id="pvalb">Pvalb</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-9.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="meis2" class="level4">
<h4 class="anchored" data-anchor-id="meis2">Meis2</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-10.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="astro" class="level4">
<h4 class="anchored" data-anchor-id="astro">Astro</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-11.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="vlmc" class="level4">
<h4 class="anchored" data-anchor-id="vlmc">VLMC</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-12.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="smc" class="level4">
<h4 class="anchored" data-anchor-id="smc">SMC</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-14-13.png" class="img-fluid" width="672"></p>
</div>
</section>
</section>
<section id="visualize-multiple-cell-types" class="level3">
<h3 class="anchored" data-anchor-id="visualize-multiple-cell-types">Visualize multiple cell types</h3>
<p>We can also visualize some of these cell types in one single plot with <code>MapMultipleFeatures()</code>.</p>
<div class="cell" data-fig.asp="0.4">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load H&amp;E images</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>se_brain_spatial <span class="ot">&lt;-</span> se_brain_spatial <span class="sc">|&gt;</span> </span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LoadImages</span>()</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot multiple features</span></span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a><span class="fu">MapMultipleFeatures</span>(se_brain_spatial, </span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">pt_size =</span> <span class="dv">2</span>, <span class="at">max_cutoff =</span> <span class="fl">0.99</span>,</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>, </span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#44AA99"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span>),</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">features =</span> selected_celltypes[<span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]) <span class="sc">+</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p><br></p>
</section>
<section id="cell-type-co-localization" class="level3">
<h3 class="anchored" data-anchor-id="cell-type-co-localization">Cell type co-localization</h3>
<p>By computing the pair-wise correlation between cell types across spots, we can get an idea of which cell types often appear together in the same spots.</p>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>cor_matrix <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(se_brain_spatial, selected_celltypes) <span class="sc">|&gt;</span> </span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_all</span>(<span class="sc">~</span> <span class="fu">if_else</span>(.x<span class="sc">&lt;</span><span class="fl">0.1</span>, <span class="dv">0</span>, .x)) <span class="sc">|&gt;</span>  <span class="co"># Filter lowest values (-&gt; set as 0)</span></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>()</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(cor_matrix) <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>max_val <span class="ot">&lt;-</span> <span class="fu">max</span>(cor_matrix, <span class="at">na.rm =</span> T)</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> RColorBrewer<span class="sc">::</span><span class="fu">brewer.pal</span>(<span class="dv">7</span>, <span class="st">"RdYlBu"</span>) <span class="sc">|&gt;</span> <span class="fu">rev</span>(); cols[<span class="dv">4</span>] <span class="ot">&lt;-</span> <span class="st">"white"</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>pheatmap<span class="sc">::</span><span class="fu">pheatmap</span>(cor_matrix, </span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">breaks =</span> <span class="fu">seq</span>(<span class="sc">-</span>max_val, max_val, <span class="at">length.out =</span> <span class="dv">100</span>),</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">color=</span><span class="fu">colorRampPalette</span>(cols)(<span class="dv">100</span>),</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>                   <span class="at">cellwidth =</span> <span class="dv">14</span>, <span class="at">cellheight =</span> <span class="dv">14</span>, </span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>                   <span class="at">treeheight_col =</span> <span class="dv">10</span>, <span class="at">treeheight_row =</span> <span class="dv">10</span>, </span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>                   <span class="at">main =</span> <span class="st">"Cell type correlation</span><span class="sc">\n</span><span class="st">within spots"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="384"></p>
</div>
</div>
<p><br></p>
<p>We can also use the predicted cell type proportions to compute factors using NMF, and thereby summarizing the cell types presence within the same location into a set of predefined factors. Each factor can be viewed as a niche of a certain cell type composition.</p>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>nmf_data <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(se_brain_spatial, selected_celltypes) <span class="sc">|&gt;</span> </span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>  RcppML<span class="sc">::</span><span class="fu">nmf</span>(<span class="at">k =</span> <span class="dv">10</span>, <span class="at">verbose =</span> F)</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>nmf_data_h <span class="ot">&lt;-</span> nmf_data<span class="sc">$</span>h <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>()</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(nmf_data_h) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">"Factor_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(nmf_data_h) <span class="ot">&lt;-</span> selected_celltypes</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>nmf_data_h <span class="ot">&lt;-</span> nmf_data_h <span class="sc">|&gt;</span> </span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate_at</span>(<span class="fu">colnames</span>(nmf_data_h), </span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>            <span class="sc">~</span>(<span class="fu">scale</span>(., <span class="at">center =</span> <span class="cn">FALSE</span>, <span class="at">scale =</span> <span class="fu">max</span>(., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)<span class="sc">/</span><span class="dv">1</span>)))</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>nmf_data_h<span class="sc">$</span>Factor <span class="ot">&lt;-</span> <span class="fu">rownames</span>(nmf_data_h) <span class="sc">|&gt;</span> </span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">factor</span>(<span class="at">levels =</span> <span class="fu">paste0</span>(<span class="st">"Factor_"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>))</span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a>nmf_data_h_df <span class="ot">&lt;-</span> nmf_data_h <span class="sc">|&gt;</span> </span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">all_of</span>(selected_celltypes), </span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a>                      <span class="at">names_to =</span> <span class="st">"Cell"</span>, </span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>                      <span class="at">values_to =</span> <span class="st">"Weight"</span>)</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(nmf_data_h_df, <span class="fu">aes</span>(<span class="at">x=</span>Factor, <span class="at">y=</span>Cell, <span class="at">size=</span>Weight, <span class="at">color=</span>Weight)) <span class="sc">+</span></span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Cell type contribution"</span>, <span class="at">x=</span><span class="st">"Factor"</span>, <span class="at">y =</span> <span class="st">"Cell type"</span>, </span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">color =</span> <span class="st">""</span>, <span class="at">size =</span> <span class="st">"Scaled weight"</span>) <span class="sc">+</span></span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_viridis_c</span>(<span class="at">direction =</span> <span class="sc">-</span><span class="dv">1</span>, <span class="at">option =</span> <span class="st">"magma"</span>) <span class="sc">+</span></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>), </span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">panel.grid =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid" width="384"></p>
</div>
</div>
</section>
</section>
<section id="mkidney" class="level2">
<h2 class="anchored" data-anchor-id="mkidney">Mouse kidney</h2>
<p>In the second example, we’ll look at data from mouse kidney. We can obtain the single-cell data with the <code>TabulaMurisSenisData</code> R package from bioconductor. Let’s load the data and create a <code>Seurat</code> object from it.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>sce <span class="ot">&lt;-</span> <span class="fu">TabulaMurisSenisDroplet</span>(<span class="at">tissues =</span> <span class="st">"Kidney"</span>)<span class="sc">$</span>Kidney</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>umis <span class="ot">&lt;-</span> <span class="fu">as</span>(<span class="fu">counts</span>(sce), <span class="st">"dgCMatrix"</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>se_kidney_singlecell <span class="ot">&lt;-</span> <span class="fu">CreateSeuratObject</span>(<span class="at">counts =</span> umis, <span class="at">meta.data =</span> <span class="fu">colData</span>(sce) <span class="sc">|&gt;</span> <span class="fu">as.data.frame</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>The 10x Visium mouse kidney data can be downloaded from the 10x genomics website.</p>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(tmpdir, <span class="st">"/kidney"</span>))</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>targetdir <span class="ot">&lt;-</span> <span class="fu">paste0</span>(tmpdir, <span class="st">"/kidney"</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium"</span>))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Download section 1</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_filtered_feature_bc_matrix.h5"</span>, </span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/filtered_feature_bc_matrix.h5"</span>))</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a><span class="fu">download.file</span>(<span class="at">url =</span> <span class="st">"https://cf.10xgenomics.com/samples/spatial-exp/1.1.0/V1_Mouse_Kidney/V1_Mouse_Kidney_spatial.tar.gz"</span>,</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>              <span class="at">destfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/spatial.tar.gz"</span>))</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a><span class="fu">untar</span>(<span class="at">tarfile =</span> <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/spatial.tar.gz"</span>), </span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">exdir =</span>  <span class="fu">paste0</span>(targetdir, <span class="st">"/visium/"</span>))</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a><span class="fu">file.remove</span>(<span class="fu">paste0</span>(targetdir, <span class="st">"/visium/spatial.tar.gz"</span>))</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="st">"./kidney/visium/filtered_feature_bc_matrix.h5"</span></span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>imgs <span class="ot">&lt;-</span> <span class="st">"./kidney/visium/spatial/tissue_hires_image.png"</span></span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>spotfiles <span class="ot">&lt;-</span> <span class="st">"./kidney/visium/spatial/tissue_positions_list.csv"</span></span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>json <span class="ot">&lt;-</span> <span class="st">"./kidney/visium/spatial/scalefactors_json.json"</span></span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>infoTable <span class="ot">&lt;-</span> tibble<span class="sc">::</span><span class="fu">tibble</span>(samples, imgs, spotfiles, json)</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>se_kidney_spatial <span class="ot">&lt;-</span> <span class="fu">ReadVisiumData</span>(infoTable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="normalize-data-1" class="level3">
<h3 class="anchored" data-anchor-id="normalize-data-1">Normalize data</h3>
<p>We apply the same normalization procedure to <code>se_kidney_spatial</code> and <code>se_kidney_singlecell</code> and run <code>FindVariableFeatures()</code> to detect the top most variable genes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>se_kidney_spatial <span class="ot">&lt;-</span> se_kidney_spatial <span class="sc">|&gt;</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For the single-cell kidney data, we’ll also filter the data prior to normalization to include cells collected at age 18m and remove cells with labels “nan” and “CD45”. This leaves us with 17 cell types.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>keep_cells <span class="ot">&lt;-</span> <span class="fu">colnames</span>(se_kidney_singlecell)[se_kidney_singlecell<span class="sc">$</span>age <span class="sc">==</span> <span class="st">"18m"</span> <span class="sc">&amp;</span> (<span class="sc">!</span>se_kidney_singlecell<span class="sc">$</span>free_annotation <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"nan"</span>, <span class="st">"CD45"</span>))]</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>se_kidney_singlecell <span class="ot">&lt;-</span> <span class="fu">subset</span>(se_kidney_singlecell, <span class="at">cells =</span> keep_cells)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>se_kidney_singlecell <span class="ot">&lt;-</span> se_kidney_singlecell <span class="sc">|&gt;</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">NormalizeData</span>() <span class="sc">|&gt;</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>() <span class="sc">|&gt;</span> </span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ScaleData</span>() <span class="sc">|&gt;</span> </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunPCA</span>() <span class="sc">|&gt;</span> </span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">RunUMAP</span>(<span class="at">reduction =</span> <span class="st">"pca"</span>, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>)</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>se_kidney_singlecell <span class="ot">&lt;-</span> se_kidney_singlecell <span class="sc">|&gt;</span> </span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">FindVariableFeatures</span>(<span class="at">nfeatures =</span> <span class="dv">10000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="run-nnls-1" class="level3">
<h3 class="anchored" data-anchor-id="run-nnls-1">Run NNLS</h3>
<p>Again, the <code>RunNNLS()</code> method requires a single-cell <code>Seurat</code> object and a 10x Visium <code>Seurat</code> object. The cell type annotations are stored in the “free_annotation” column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>ti <span class="ot">&lt;-</span> <span class="fu">Sys.time</span>()</span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_kidney_spatial) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>se_kidney_spatial <span class="ot">&lt;-</span> <span class="fu">RunNNLS</span>(<span class="at">object =</span> se_kidney_spatial, </span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">singlecell_object =</span> se_kidney_singlecell, </span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">groups =</span> <span class="st">"free_annotation"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Predicting cell type proportions ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code></code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Fetching data from Seurat objects</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Filtering out features that are only present in one data set</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Kept 4993 features for deconvolution</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Preparing data for NNLS</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Downsampling scRNA-seq data to include a maximum of 50 cells per cell type</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Cell type(s) CD45    NK cell,CD45    plasma cell have too few cells (&lt;10) and will be excluded</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Kept 15 cell types after filtering</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>→   Calculating cell type expression profiles</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Predicting cell type proportions with NNLS for 15 cell types</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Returning results in a new 'Assay' named 'celltypeprops'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ℹ Setting default assay to 'celltypeprops'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ Finished</code></pre>
</div>
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sprintf</span>(<span class="st">"RunNNLS finished in %s seconds"</span>, <span class="fu">round</span>(<span class="fu">Sys.time</span>() <span class="sc">-</span> ti, <span class="at">digits =</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "RunNNLS finished in 0.39 seconds"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check available cell types</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(se_kidney_spatial)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "CD45    B cell"                                            
 [2] "CD45    T cell"                                            
 [3] "CD45    macrophage"                                        
 [4] "Epcam     kidney distal convoluted tubule epithelial cell" 
 [5] "Epcam    brush cell"                                       
 [6] "Epcam    kidney collecting duct principal cell"            
 [7] "Epcam    kidney proximal convoluted tubule epithelial cell"
 [8] "Epcam    podocyte"                                         
 [9] "Epcam    proximal tube epithelial cell"                    
[10] "Epcam    thick ascending tube S epithelial cell"           
[11] "Pecam    Kidney cortex artery cell"                        
[12] "Pecam    fenestrated capillary endothelial"                
[13] "Pecam    kidney capillary endothelial cell"                
[14] "Stroma    fibroblast"                                      
[15] "Stroma    kidney mesangial cell"                           </code></pre>
</div>
</div>
<p>NB: two cell types were discarded as they didn’t pass the minimum allowed cells per cell type threshold. The cell types discarded were NK cells and plasma cells.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r fold-hide code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot selected cell types</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(se_kidney_spatial) <span class="ot">&lt;-</span> <span class="st">"celltypeprops"</span></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>selected_celltypes <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Epcam     kidney distal convoluted tubule epithelial cell"</span>,</span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    brush cell"</span>,</span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    kidney collecting duct principal cell"</span>,</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    kidney proximal convoluted tubule epithelial cell"</span>,</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    podocyte"</span>,</span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    proximal tube epithelial cell"</span>,</span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Epcam    thick ascending tube S epithelial cell"</span>,</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Pecam    fenestrated capillary endothelial"</span>,</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Pecam    kidney capillary endothelial cell"</span>,</span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>                        <span class="st">"Stroma    kidney mesangial cell"</span>)</span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a>plots <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">seq_along</span>(selected_celltypes), <span class="cf">function</span>(i) {</span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">MapFeatures</span>(se_kidney_spatial, <span class="at">pt_size =</span> <span class="fl">1.3</span>,</span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a>            <span class="at">features =</span> selected_celltypes[i], </span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a>            <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> </span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">setNames</span>(<span class="at">nm =</span> selected_celltypes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="mapped-kidney-cell-types" class="level3 tabset">
<h3 class="tabset anchored" data-anchor-id="mapped-kidney-cell-types">Mapped kidney cell types</h3>
<section id="epcam-kidney-distal-convoluted-tubule-epithelial-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-kidney-distal-convoluted-tubule-epithelial-cell">Epcam kidney distal convoluted tubule epithelial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-1.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-brush-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-brush-cell">Epcam brush cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-2.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-kidney-collecting-duct-principal-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-kidney-collecting-duct-principal-cell">Epcam kidney collecting duct principal cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-3.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-kidney-proximal-convoluted-tubule-epithelial-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-kidney-proximal-convoluted-tubule-epithelial-cell">Epcam kidney proximal convoluted tubule epithelial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-4.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-podocyte" class="level4">
<h4 class="anchored" data-anchor-id="epcam-podocyte">Epcam podocyte</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-5.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-proximal-tube-epithelial-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-proximal-tube-epithelial-cell">Epcam proximal tube epithelial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-6.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="epcam-thick-ascending-tube-s-epithelial-cell" class="level4">
<h4 class="anchored" data-anchor-id="epcam-thick-ascending-tube-s-epithelial-cell">Epcam thick ascending tube S epithelial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-7.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="pecam-fenestrated-capillary-endothelial" class="level4">
<h4 class="anchored" data-anchor-id="pecam-fenestrated-capillary-endothelial">Pecam fenestrated capillary endothelial</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-8.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="pecam-kidney-capillary-endothelial-cell" class="level4">
<h4 class="anchored" data-anchor-id="pecam-kidney-capillary-endothelial-cell">Pecam kidney capillary endothelial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-9.png" class="img-fluid" width="672"></p>
</div>
</section>
<section id="stroma-kidney-mesangial-cell" class="level4">
<h4 class="anchored" data-anchor-id="stroma-kidney-mesangial-cell">Stroma kidney mesangial cell</h4>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-26-10.png" class="img-fluid" width="672"></p>
</div>
</section>
</section>
<section id="visualize-multiple-cell-types-1" class="level3">
<h3 class="anchored" data-anchor-id="visualize-multiple-cell-types-1">Visualize multiple cell types</h3>
<p>We can select some of these cell types and visualize them in one plot</p>
<div class="cell" data-fig.asp="0.4">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load H&amp;E images</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>se_kidney_spatial <span class="ot">&lt;-</span> se_kidney_spatial <span class="sc">|&gt;</span> </span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">LoadImages</span>()</span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot multiple features</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a><span class="fu">MapMultipleFeatures</span>(se_kidney_spatial, </span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">pt_size =</span> <span class="dv">2</span>, <span class="at">max_cutoff =</span> <span class="fl">0.95</span>,</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>, </span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span>),</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>                    <span class="at">features =</span> selected_celltypes[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">7</span>)]) <span class="sc">+</span></span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p>Or zoom in on a region of interest</p>
<div class="cell" data-fig.asp="0.8">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Reload H&amp;E image in higher resolution</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>se_kidney_spatial <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(se_kidney_spatial, <span class="at">image_height =</span> <span class="dv">1500</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot H&amp;E image</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>rst <span class="ot">&lt;-</span> <span class="fu">GetImages</span>(se_kidney_spatial)[[<span class="dv">1</span>]]</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>()  <span class="sc">+</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"H&amp;E image of zoomed in area"</span>) <span class="sc">+</span> </span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.2</span>)) <span class="sc">+</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">inset_element</span>(<span class="at">p =</span> rst[(<span class="fl">0.5</span><span class="sc">*</span><span class="fu">nrow</span>(rst))<span class="sc">:</span>(<span class="fl">0.7</span><span class="sc">*</span><span class="fu">nrow</span>(rst)), (<span class="fl">0.5</span><span class="sc">*</span><span class="fu">ncol</span>(rst))<span class="sc">:</span>(<span class="fl">0.7</span><span class="sc">*</span><span class="fu">ncol</span>(rst))], </span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>                <span class="at">left =</span> <span class="dv">0</span>, <span class="at">bottom =</span> <span class="dv">0</span>, <span class="at">right =</span> <span class="dv">1</span>, <span class="at">top =</span> <span class="dv">1</span>)</span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot multiple features with zoom</span></span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapMultipleFeatures</span>(se_kidney_spatial, </span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a>                    <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a>                    <span class="at">pt_size =</span> <span class="fl">4.5</span>, <span class="at">max_cutoff =</span> <span class="fl">0.95</span>,</span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a>                    <span class="at">crop_area =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.5</span>, <span class="fl">0.7</span>, <span class="fl">0.7</span>),</span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a>                    <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"#332288"</span>, <span class="st">"#88CCEE"</span>, <span class="st">"#117733"</span>, <span class="st">"#DDCC77"</span>, <span class="st">"#CC6677"</span>,<span class="st">"#AA4499"</span>),</span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a>                    <span class="at">features =</span> selected_celltypes[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>, <span class="dv">6</span>, <span class="dv">7</span>)]) <span class="sc">+</span></span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot_layout</span>(<span class="at">guides =</span> <span class="st">"collect"</span>) <span class="sc">&amp;</span></span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">/</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="cell_type_mapping_with_NNLS_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<p><br></p>
<hr>
<details open="">
<summary>
<strong>Package versions</strong>
</summary>
<ul>
<li><p><code>semla</code>: 1.1.4</p></li>
<li><p><code>RcppML</code>: 0.3.7</p></li>
</ul>
</details>
<details>
<summary>
<strong>Session info</strong>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.3 (2023-03-15)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] patchwork_1.1.3             purrr_1.0.2                
 [3] SingleCellExperiment_1.18.1 SummarizedExperiment_1.26.1
 [5] Biobase_2.56.0              GenomicRanges_1.48.0       
 [7] GenomeInfoDb_1.32.4         IRanges_2.30.1             
 [9] S4Vectors_0.34.0            BiocGenerics_0.42.0        
[11] MatrixGenerics_1.8.1        matrixStats_1.0.0          
[13] TabulaMurisSenisData_1.2.0  tibble_3.2.1               
[15] semla_1.1.4                 ggplot2_3.4.3              
[17] dplyr_1.1.2                 SeuratObject_4.1.3         
[19] Seurat_4.3.0.1             

loaded via a namespace (and not attached):
  [1] utf8_1.2.3                    spatstat.explore_3.2-1       
  [3] reticulate_1.31               tidyselect_1.2.0             
  [5] RSQLite_2.3.1                 AnnotationDbi_1.58.0         
  [7] htmlwidgets_1.6.2             grid_4.2.3                   
  [9] Rtsne_0.16                    munsell_0.5.0                
 [11] codetools_0.2-19              ica_1.0-3                    
 [13] future_1.33.0                 miniUI_0.1.1.1               
 [15] withr_2.5.0                   spatstat.random_3.1-5        
 [17] colorspace_2.1-0              progressr_0.14.0             
 [19] filelock_1.0.2                knitr_1.43                   
 [21] rstudioapi_0.15.0             ROCR_1.0-11                  
 [23] tensor_1.5                    listenv_0.9.0                
 [25] labeling_0.4.3                GenomeInfoDbData_1.2.8       
 [27] polyclip_1.10-4               RcppML_0.3.7                 
 [29] bit64_4.0.5                   farver_2.1.1                 
 [31] pheatmap_1.0.12               rhdf5_2.40.0                 
 [33] parallelly_1.36.0             vctrs_0.6.3                  
 [35] generics_0.1.3                xfun_0.40                    
 [37] BiocFileCache_2.4.0           R6_2.5.1                     
 [39] bitops_1.0-7                  rhdf5filters_1.8.0           
 [41] spatstat.utils_3.0-3          cachem_1.0.8                 
 [43] DelayedArray_0.22.0           promises_1.2.1               
 [45] scales_1.2.1                  gtable_0.3.4                 
 [47] globals_0.16.2                goftest_1.2-3                
 [49] rlang_1.1.1                   zeallot_0.1.0                
 [51] splines_4.2.3                 lazyeval_0.2.2               
 [53] spatstat.geom_3.2-4           BiocManager_1.30.22          
 [55] yaml_2.3.7                    reshape2_1.4.4               
 [57] abind_1.4-5                   httpuv_1.6.11                
 [59] tools_4.2.3                   ellipsis_0.3.2               
 [61] RColorBrewer_1.1-3            ggridges_0.5.4               
 [63] Rcpp_1.0.11                   plyr_1.8.8                   
 [65] zlibbioc_1.42.0               RCurl_1.98-1.12              
 [67] dbscan_1.1-11                 deldir_1.0-9                 
 [69] pbapply_1.7-2                 cowplot_1.1.1                
 [71] zoo_1.8-12                    ggrepel_0.9.3                
 [73] cluster_2.1.4                 magrittr_2.0.3               
 [75] data.table_1.14.8             magick_2.7.5                 
 [77] scattermore_1.2               lmtest_0.9-40                
 [79] RANN_2.6.1                    ggnewscale_0.4.9             
 [81] fitdistrplus_1.1-11           shinyjs_2.1.0                
 [83] mime_0.12                     evaluate_0.21                
 [85] xtable_1.8-4                  gridExtra_2.3                
 [87] compiler_4.2.3                KernSmooth_2.23-22           
 [89] crayon_1.5.2                  htmltools_0.5.6              
 [91] later_1.3.1                   tidyr_1.3.0                  
 [93] DBI_1.1.3                     ExperimentHub_2.4.0          
 [95] dbplyr_2.3.3                  MASS_7.3-60                  
 [97] rappdirs_0.3.3                Matrix_1.6-1                 
 [99] cli_3.6.1                     gdata_2.19.0                 
[101] parallel_4.2.3                igraph_1.5.1                 
[103] forcats_1.0.0                 pkgconfig_2.0.3              
[105] sp_2.0-0                      plotly_4.10.2                
[107] spatstat.sparse_3.0-2         XVector_0.36.0               
[109] stringr_1.5.0                 digest_0.6.33                
[111] sctransform_0.3.5             RcppAnnoy_0.0.21             
[113] spatstat.data_3.0-1           Biostrings_2.64.1            
[115] rmarkdown_2.24                leiden_0.4.3                 
[117] uwot_0.1.16                   curl_5.0.2                   
[119] shiny_1.7.5                   gtools_3.9.4                 
[121] lifecycle_1.0.3               nlme_3.1-163                 
[123] jsonlite_1.8.7                Rhdf5lib_1.18.2              
[125] viridisLite_0.4.2             fansi_1.0.4                  
[127] pillar_1.9.0                  lattice_0.21-8               
[129] KEGGREST_1.36.3               fastmap_1.1.1                
[131] httr_1.4.7                    survival_3.5-7               
[133] interactiveDisplayBase_1.34.0 glue_1.6.2                   
[135] png_0.1-8                     BiocVersion_3.15.2           
[137] bit_4.0.5                     stringi_1.7.12               
[139] HDF5Array_1.24.2              blob_1.2.4                   
[141] AnnotationHub_3.4.0           memoise_2.0.1                
[143] irlba_2.3.5.1                 future.apply_1.11.0          </code></pre>
</div>
</div>
</details>
<p><br></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>