<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-0.9.649">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-08-31">

<title>Visium + Immunofluorescence data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="IF_data_files/libs/clipboard/clipboard.min.js"></script>
<script src="IF_data_files/libs/quarto-html/quarto.js"></script>
<script src="IF_data_files/libs/quarto-html/popper.min.js"></script>
<script src="IF_data_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="IF_data_files/libs/quarto-html/anchor.min.js"></script>
<link href="IF_data_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="IF_data_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="IF_data_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="IF_data_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="IF_data_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Visium + Immunofluorescence data</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 31, 2023</p>
    </div>
  </div>
    
  </div>
  

</header>

<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(semla)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In this article, we’ll demonstrate how to load Immunoflurescence (IF) data with <code>semla</code>.</p>
<p>You can download example data from 10x Genomics website. Here we’ll use a Breast Cancer data set called “Visium CytAssist Gene and Protein Expression Library of Human Breast Cancer, IF, 6.5mm (FFPE)”.</p>
<p>Assuming that you have downloaded the data (filtered_feature_bc_matrix.h5 and the spatial/ folder) into your current working directory, you can load the data into a Seurat object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span> <span class="st">"filtered_feature_bc_matrix.h5"</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>imgs <span class="ot">&lt;-</span> <span class="st">"spatial/tissue_hires_image.png"</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>spotfiles <span class="ot">&lt;-</span> <span class="st">"spatial/tissue_positions.csv"</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>json <span class="ot">&lt;-</span> <span class="st">"spatial/scalefactors_json.json"</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>infoTable <span class="ot">&lt;-</span> <span class="fu">tibble</span>(samples, imgs, spotfiles, json)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">ReadVisiumData</span>(infoTable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we have access to both gene expression counts stored in the ‘Spatial’ assay and antibody capture measurements stored in the ‘AbCapture’ assay.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>hBrCa</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
18120 features across 4169 samples within 2 assays 
Active assay: Spatial (18085 features, 0 variable features)
 1 other assay present: AbCapture</code></pre>
</div>
</div>
<p>This data set comes with a immunofluorescence image that we can use as a background for our plots.</p>
<div class="cell" data-fig.asp="0.98">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(hBrCa, <span class="at">image_height =</span> <span class="dv">1900</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ImagePlot</span>(hBrCa, <span class="at">mar =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/raw_image.jpg" class="img-fluid"></p>
<p>Now we can plot both RNA expression values and antibody capture measurements with <code>semla</code>. But first, let’s normalize the expression values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> hBrCa <span class="sc">|&gt;</span> <span class="fu">NormalizeData</span>()</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> hBrCa <span class="sc">|&gt;</span> <span class="fu">NormalizeData</span>(<span class="at">normalization.method =</span> <span class="st">"CLR"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Plot expression values:</p>
<div class="cell" data-fig.asp="0.5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1"</span>, <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p1, <span class="at">titles =</span> <span class="st">"Gene expression (normalized)"</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1.1"</span>, <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> </span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p2, <span class="at">titles =</span> <span class="st">"Protein expression (normalized)"</span>)</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/MS4A1_no_HE.jpg" class="img-fluid"></p>
<p>If we include the immunofluorescence image as the background for our plots, we might want to modify the plots to have a black theme.</p>
<div class="cell" data-fig.asp="0.5">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>dark_theme <span class="ot">&lt;-</span> <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">"black"</span>, <span class="at">color =</span> <span class="st">"black"</span>),</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                    <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"white"</span>),</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                    <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"white"</span>),</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"white"</span>),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">colour =</span> <span class="st">"white"</span>))</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1"</span>, <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>, </span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> viridis<span class="sc">::</span><span class="fu">viridis</span>(<span class="at">n =</span> <span class="dv">11</span>), <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> </span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>() <span class="sc">&amp;</span> dark_theme</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p1, <span class="at">titles =</span> <span class="st">"Gene expression"</span>)</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1.1"</span>, <span class="at">override_plot_dims =</span> <span class="cn">TRUE</span>, </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> viridis<span class="sc">::</span><span class="fu">viridis</span>(<span class="at">n =</span> <span class="dv">11</span>), <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>) <span class="sc">&amp;</span> </span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>() <span class="sc">&amp;</span> dark_theme</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p2, <span class="at">titles =</span> <span class="st">"Protein expression"</span>)</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/MS4A1_IF_bg.jpg" class="img-fluid"></p>
<p>With this global visualization, it’s quite difficult to see the immunofluorescence image in the background. If we zoom in on a smaller area, it become slightly easier to compare the expression values with the background image.</p>
<p>The plots below show a zoom in view of the bottom right corner.</p>
<div class="cell" data-fig.asp="0.7">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1"</span>, </span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">crop_area =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.65</span>, <span class="fl">0.85</span>, <span class="dv">1</span>), <span class="at">pt_size =</span> <span class="dv">2</span>,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> viridis<span class="sc">::</span><span class="fu">viridis</span>(<span class="at">n =</span> <span class="dv">11</span>)) <span class="sc">&amp;</span> </span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>() <span class="sc">&amp;</span> dark_theme</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p1, <span class="at">titles =</span> <span class="st">"Gene expression"</span>)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"MS4A1.1"</span>, </span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">crop_area =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.65</span>, <span class="fl">0.85</span>, <span class="dv">1</span>), <span class="at">pt_size =</span> <span class="dv">2</span>, </span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> viridis<span class="sc">::</span><span class="fu">viridis</span>(<span class="at">n =</span> <span class="dv">11</span>)) <span class="sc">&amp;</span> </span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>() <span class="sc">&amp;</span> dark_theme</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(<span class="at">p =</span> p2, <span class="at">titles =</span> <span class="st">"Protein expression"</span>)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ImagePlot</span>(hBrCa, <span class="at">crop_area =</span> <span class="fu">c</span>(<span class="fl">0.5</span>, <span class="fl">0.65</span>, <span class="fl">0.85</span>, <span class="dv">1</span>), <span class="at">return_as_gg =</span> <span class="cn">TRUE</span>)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>p3 <span class="sc">+</span> p1 <span class="sc">+</span> p2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">design =</span> <span class="fu">c</span>(<span class="fu">area</span>(<span class="at">t =</span> <span class="dv">1</span>, <span class="at">l =</span> <span class="dv">1</span>, <span class="at">b =</span> <span class="dv">2</span>, <span class="at">r =</span> <span class="dv">2</span>),</span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">area</span>(<span class="at">t =</span> <span class="dv">1</span>, <span class="at">l =</span> <span class="dv">3</span>, <span class="at">b =</span> <span class="dv">1</span>, <span class="at">r =</span> <span class="dv">3</span>),</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">area</span>(<span class="at">t =</span> <span class="dv">2</span>, <span class="at">l =</span> <span class="dv">3</span>, <span class="at">b =</span> <span class="dv">2</span>, <span class="at">r =</span> <span class="dv">3</span>))) <span class="sc">&amp;</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  dark_theme</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/MS4A1_IF_bg_zoom.jpg" class="img-fluid"></p>
<section id="change-background-image-for-visualization" class="level1">
<h1>Change background image for visualization</h1>
<p>We can also easily switch the images in <code>semla</code>. The only requirement is that the alternative image is aligned with the spots and has the same aspect ratio as the original image used for Space Ranger.</p>
<p>The code below illustrated how one can select the red and green color channels from the immunofluorescence image and reload them for visualization with <code>semla</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(magick)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Fetch the path to the immunofluoresence (IF) image</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>impath <span class="ot">&lt;-</span> <span class="fu">GetStaffli</span>(hBrCa)<span class="sc">@</span>imgs</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Load IF image</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>im <span class="ot">&lt;-</span> <span class="fu">image_read</span>(<span class="at">path =</span> impath)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract channels. Here we only select the red and green channels, </span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># but we could of course also use the DAPI stain in the blue channel</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>cols <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>)</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>channels <span class="ot">&lt;-</span> <span class="fu">lapply</span>(cols, <span class="cf">function</span>(col) {</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_channel</span>(im, <span class="at">channel =</span> col)</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>}) <span class="sc">|&gt;</span> <span class="fu">setNames</span>(<span class="at">nm =</span> cols)</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply simple background correction to reduce autofluorescence</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Note that this code is just for demonstration. </span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>channels <span class="ot">&lt;-</span> <span class="fu">lapply</span>(channels, <span class="cf">function</span>(ch) {</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>  im_blurred <span class="ot">&lt;-</span> <span class="fu">image_median</span>(ch, <span class="at">radius =</span> <span class="dv">20</span>)</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>  im_composite <span class="ot">&lt;-</span> <span class="fu">image_composite</span>(im_blurred, ch, <span class="at">operator =</span> <span class="st">"Minus"</span>)</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(im_composite)</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Export images</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>channels<span class="sc">$</span>green <span class="sc">|&gt;</span> </span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_normalize</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_write</span>(<span class="at">path =</span> <span class="st">"IF_data/normalized_image_green.png"</span>, </span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>              <span class="at">format =</span> <span class="st">"png"</span>, <span class="at">quality =</span> <span class="dv">100</span>)</span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>channels<span class="sc">$</span>red <span class="sc">|&gt;</span> </span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_normalize</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_write</span>(<span class="at">path =</span> <span class="st">"IF_data/normalized_image_red.png"</span>, </span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>              <span class="at">format =</span> <span class="st">"png"</span>, <span class="at">quality =</span> <span class="dv">100</span>)</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>im <span class="sc">|&gt;</span> </span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_normalize</span>() <span class="sc">|&gt;</span> </span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>  <span class="fu">image_write</span>(<span class="at">path =</span> <span class="st">"IF_data/normalized_image.png"</span>, </span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>              <span class="at">format =</span> <span class="st">"png"</span>, <span class="at">quality =</span> <span class="dv">100</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="use-green-channel-pcna-as-background" class="level2">
<h2 class="anchored" data-anchor-id="use-green-channel-pcna-as-background">Use green channel (PCNA) as background</h2>
<p>In order to switch the image in the Seurat object, we can just replace the path to the image using <code>ReplaceImagePaths</code> and then reload the image to desired size with <code>LoadImages</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">ReplaceImagePaths</span>(hBrCa, <span class="at">paths =</span> <span class="st">"IF_data/normalized_image_green.png"</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(hBrCa, <span class="at">image_height =</span> <span class="dv">1957</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can make plots with <code>semla</code> using the new image as background.</p>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"PCNA.1"</span>, <span class="at">image_use =</span> <span class="st">"raw"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>, </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>, <span class="at">pt_size =</span> <span class="fl">1.5</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"lightgrey"</span>, <span class="st">"orange"</span>), <span class="at">max_cutoff =</span> <span class="fl">0.99</span>) <span class="sc">&amp;</span> </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p1, <span class="at">titles =</span> <span class="st">"Protein expression (raw values)"</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"PCNA"</span>, <span class="at">image_use =</span> <span class="st">"raw"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>, </span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>, <span class="at">pt_size =</span> <span class="fl">1.5</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"lightgrey"</span>, <span class="st">"orange"</span>), <span class="at">max_cutoff =</span> <span class="fl">0.99</span>) <span class="sc">&amp;</span> </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p2, <span class="at">titles =</span> <span class="st">"RNA expression (UMI counts)"</span>)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ImagePlot</span>(hBrCa, <span class="at">return_as_gg =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p3, <span class="at">titles =</span> <span class="st">"IF signal (corrected)"</span>)</span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> p3 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">&amp;</span> dark_theme</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/PCNA_composite.jpg" class="img-fluid"></p>
</section>
<section id="use-red-channel-vim-as-background" class="level2">
<h2 class="anchored" data-anchor-id="use-red-channel-vim-as-background">Use red channel (VIM) as background</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">ReplaceImagePaths</span>(hBrCa, <span class="at">paths =</span> <span class="st">"IF_data/normalized_image_red.png"</span>)</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(hBrCa, <span class="at">image_height =</span> <span class="dv">1957</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-fig.asp="1">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"VIM.1"</span>, <span class="at">image_use =</span> <span class="st">"raw"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>, </span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>, <span class="at">pt_size =</span> <span class="fl">1.5</span>,</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"lightgrey"</span>, <span class="st">"orange"</span>), <span class="at">max_cutoff =</span> <span class="fl">0.99</span>) <span class="sc">&amp;</span> </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p1, <span class="at">titles =</span> <span class="st">"Protein expression (raw values)"</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"Spatial"</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="st">"VIM"</span>, <span class="at">image_use =</span> <span class="st">"raw"</span>, <span class="at">slot =</span> <span class="st">"counts"</span>, </span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>                  <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>, <span class="at">pt_size =</span> <span class="fl">1.5</span>,</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"lightgrey"</span>, <span class="st">"orange"</span>), <span class="at">max_cutoff =</span> <span class="fl">0.99</span>) <span class="sc">&amp;</span> </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>()</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p2, <span class="at">titles =</span> <span class="st">"RNA expression (UMI counts)"</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ImagePlot</span>(hBrCa, <span class="at">return_as_gg =</span> <span class="cn">TRUE</span>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">ModifyPatchworkTitles</span>(p3, <span class="at">titles =</span> <span class="st">"IF signal (corrected)"</span>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2 <span class="sc">+</span> p3 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>) <span class="sc">&amp;</span> dark_theme</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/VIM_composite.jpg" class="img-fluid"></p>
</section>
<section id="use-normalized-if-image-as-background" class="level2">
<h2 class="anchored" data-anchor-id="use-normalized-if-image-as-background">Use normalized IF image as background</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">ReplaceImagePaths</span>(hBrCa, <span class="at">paths =</span> <span class="st">"IF_data/normalized_image.png"</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hBrCa <span class="ot">&lt;-</span> <span class="fu">LoadImages</span>(hBrCa, <span class="at">image_height =</span> <span class="dv">1957</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ImgePlot</span>(hBrCa)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/norm_IF.jpg" class="img-fluid"></p>
<div class="cell" data-fig.asp="0.5">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(hBrCa) <span class="ot">&lt;-</span> <span class="st">"AbCapture"</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">MapFeatures</span>(hBrCa, <span class="at">features =</span> <span class="fu">c</span>(<span class="st">"PCNA.1"</span>, <span class="st">"VIM.1"</span>), <span class="at">image_use =</span> <span class="st">"raw"</span>, </span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">slot =</span> <span class="st">"counts"</span>, <span class="at">colors =</span> <span class="fu">c</span>(<span class="st">"lightgrey"</span>, <span class="st">"yellow"</span>), <span class="at">pt_size =</span> <span class="fl">1.7</span>, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">scale_alpha =</span> <span class="cn">TRUE</span>, <span class="at">max_cutoff =</span> <span class="fl">0.99</span>) <span class="sc">&amp;</span> </span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ThemeLegendRight</span>() <span class="sc">&amp;</span> dark_theme <span class="sc">&amp;</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><img src="IF_data/norm_IF_and_expr.jpg" class="img-fluid"></p>
<p><br></p>
<hr>
<details open="">
<summary>
<strong>Package versions</strong>
</summary>
<ul>
<li><p><code>semla</code>: 1.1.4</p></li>
<li><p><code>RcppML</code>: 0.3.7</p></li>
</ul>
</details>
<details>
<summary>
<strong>Session info</strong>
</summary>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>R version 4.2.3 (2023-03-15)
Platform: x86_64-apple-darwin13.4.0 (64-bit)
Running under: macOS Big Sur ... 10.16

Matrix products: default
BLAS/LAPACK: /Users/ludviglarsson/miniconda3/envs/R4.2/lib/libopenblasp-r0.3.21.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] semla_1.1.4        ggplot2_3.4.3      dplyr_1.1.2        SeuratObject_4.1.3
[5] Seurat_4.3.0.1    

loaded via a namespace (and not attached):
  [1] Rtsne_0.16             colorspace_2.1-0       deldir_1.0-9          
  [4] ellipsis_0.3.2         ggridges_0.5.4         rstudioapi_0.15.0     
  [7] spatstat.data_3.0-1    leiden_0.4.3           listenv_0.9.0         
 [10] ggrepel_0.9.3          fansi_1.0.4            codetools_0.2-19      
 [13] splines_4.2.3          knitr_1.43             zeallot_0.1.0         
 [16] polyclip_1.10-4        jsonlite_1.8.7         ica_1.0-3             
 [19] cluster_2.1.4          png_0.1-8              uwot_0.1.16           
 [22] shiny_1.7.5            sctransform_0.3.5      spatstat.sparse_3.0-2 
 [25] compiler_4.2.3         httr_1.4.7             Matrix_1.6-1          
 [28] fastmap_1.1.1          lazyeval_0.2.2         cli_3.6.1             
 [31] later_1.3.1            htmltools_0.5.6        tools_4.2.3           
 [34] igraph_1.5.1           gtable_0.3.4           glue_1.6.2            
 [37] RANN_2.6.1             reshape2_1.4.4         Rcpp_1.0.11           
 [40] scattermore_1.2        vctrs_0.6.3            spatstat.explore_3.2-1
 [43] nlme_3.1-163           progressr_0.14.0       lmtest_0.9-40         
 [46] spatstat.random_3.1-5  xfun_0.40              stringr_1.5.0         
 [49] globals_0.16.2         mime_0.12              miniUI_0.1.1.1        
 [52] lifecycle_1.0.3        irlba_2.3.5.1          goftest_1.2-3         
 [55] future_1.33.0          MASS_7.3-60            zoo_1.8-12            
 [58] scales_1.2.1           promises_1.2.1         spatstat.utils_3.0-3  
 [61] parallel_4.2.3         RColorBrewer_1.1-3     yaml_2.3.7            
 [64] reticulate_1.31        pbapply_1.7-2          gridExtra_2.3         
 [67] stringi_1.7.12         rlang_1.1.1            pkgconfig_2.0.3       
 [70] matrixStats_1.0.0      evaluate_0.21          lattice_0.21-8        
 [73] ROCR_1.0-11            purrr_1.0.2            tensor_1.5            
 [76] patchwork_1.1.3        htmlwidgets_1.6.2      cowplot_1.1.1         
 [79] tidyselect_1.2.0       parallelly_1.36.0      RcppAnnoy_0.0.21      
 [82] plyr_1.8.8             magrittr_2.0.3         R6_2.5.1              
 [85] magick_2.7.5           generics_0.1.3         pillar_1.9.0          
 [88] withr_2.5.0            fitdistrplus_1.1-11    survival_3.5-7        
 [91] abind_1.4-5            sp_2.0-0               tibble_3.2.1          
 [94] future.apply_1.11.0    KernSmooth_2.23-22     utf8_1.2.3            
 [97] spatstat.geom_3.2-4    plotly_4.10.2          rmarkdown_2.24        
[100] grid_4.2.3             data.table_1.14.8      forcats_1.0.0         
[103] digest_0.6.33          xtable_1.8-4           dbscan_1.1-11         
[106] tidyr_1.3.0            httpuv_1.6.11          munsell_0.5.0         
[109] viridisLite_0.4.2      shinyjs_2.1.0         </code></pre>
</div>
</div>
</details>
<p><br></p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>